onebusaway-android/src/main/java/org/onebusaway/android/tripservice/TripService.java

```diff
@@ -111,33 +111,34 @@ public TaskContextImpl(int startId) {
         public void taskComplete() {
             //Log.d(TAG, "Task complete: " + mStartId);
             // If we have notifications, then we can't stop ourselves.
-            if (mNotifications.isEmpty()) {
+            //TODO: Is there actually a reason to not stop ourselves if we are no longer storing notifications?
```

Comment:
>I am not sure that keeping the object as is would work either, since we no
longer locally store notifications.

@charles-b-stb It doesn't look like persistence of notification records is changed at all in this PR.  Persistence to database is implemented in `SchedulerTask.scheduleAlert()`.  I believe this is ok, though - we still need a persisted record of notifications so they can be re-registered with Android platform after a device power cycle (via our `BootstrapService`, which is a `BroadcastReceiver` that listens for a boot broadcast from the platform).

> (and its now unsafe to do so).  

I think persisting notifications, and using those IDs to create new notifications, should still be safe.  We just don't want to reuse `Notification` objects.  From my conversation [here](https://code.google.com/p/android/issues/detail?id=73720#c5) with Google on the best practice for updating notifications:

**Me:**

>To clarify the above, for people fixing old code - the current official best practice for Notifications is *NOT* to hold a reference and update them with new information, correct?  Instead, to show new info to the user, you should always cancel the existing notification and build a new Notification via Notification.Builder?

>The app I was working on had periodic notifications delivered to the user, and it used persisted notification IDs along with setLatestEventInfo() to update the information shown to the user within the same Notification.  Based on my understanding of the above, the proper behavior would be to cancel and re-create the Notification using the Builder.

**Google:**

>There is no need to cancel. Simply notify again with the same id and tag and the existing notification will be updated.  The actual notification object is parceled when it is sent to the service, so it's not important to hold a reference to the local Notification object.

>See packages/experimental/NotificationShowcase and look at the upload notification for an example.

The key being "Simply notify again with the same id and tag and the existing notification will be updated."  Let me know if you interpret this differently.

>The list would always be empty in every case.

I'm suggesting that we restore `mNotifications` data structure to be used much like it was before - so it would continue to be a container for `Notification` objects, but we wouldn't be re-using the `Notification` objects for actually setting the notifications - `TaskContenxt.getNotification()` would still be removed.  Instead, `mNotifications` just served as a synchronized structure for us to tell whether or not it's safe to shut down the service.

>I think the key to shutting down safely (if there is
one) is in something like what is outlined here:  http://stackoverflow.com/questions/3269445/executorservice-how-to-wait-for-all-tasks-to-finish

This is what I previously thought - my solution proposed in my previous comment with the `Future` objects is pretty close to the first solution proposed in the above SO post.  The main problem I see here with the current design is that we're trying to monitor the status of the thread pool from one of the threads in the pool.  To get around this we'd actually need to create another thread outside the pool that watchdogs the pool.  This seems messy and prone to things going wrong, especially since that SO post doesn't reference Android at all (and Android gets kill-happy when it's pressed for memory).

>I think the best way to test this is to save a reminder (Reminder A). Reminder A will trigger the PollerTask to start executing 30 min ahead of the scheduled arrival time. While Reminder A's PollerTask is executing, schedule a new reminder (Reminder B), which will start a SchedulerTask in the same service. Reminder A's PollerTask should still be executing after the SchedulerTask for Reminder B completes.

@cagryInside After looking at the code more, I believe my comment above is wrong - the `PollerTask` actually reschedules itself each time it polls (in `poll1()` line 108), so the `TripService` service could shut down in between polling events.  That's going to make this hard to test.

At this point, I'm inclined to pursue the least invasive option, as it is going to be hard to test that everything is working exactly as it did before.  This would be replacing the `setLatestInfo()` method with the `createNotification()`, so we're always creating a new notification, but using the same ID as before.  That way if the notification is still active Android should just update it.  The use of `mNotifications` would otherwise remain the same as before.

@charles-b-stb Could you please update this PR with those changes?  I think it's largely just uncommenting code that you currently have commented out.  I'd suggest saving a local copy of this commit just in case we figure out that we want to go back to this solution.  You should also remove `TaskContext.getNotification()` and it's implementation, as we shouldn't need that any longer.  And please clean up any remaining TODOs in comments.

Thanks again for working on this!  The reminders feature isn't trivial to modify :).